### PR TITLE
Fixes an issue with hosts having an IPv6 address on localhost

### PR DESCRIPTION
**Pull request by akram**
- When 'getent hosts localhost' returns '::1' the creation of the listener fails because of the port parsing which uses ":" as a separator
- Use of net.SplitHostPort() to do the job
- Adding unit tests to ensure that the creation succeeds
- On docker.go: adds a test on the presence the socat command which was failing silenty if not installed

### CODE REVIEW

**Comment by akram** [comment]
Hi guys,

this PR fixes an issue on host that have its /etc/hosts having a ::1 address mapping to localhost, and on which name resolution to localhost first resolves to the IPv6 address, which seems to be the case for fedora and arch hosts. Probably more.

To check if it the the case for your distribution run the following command:
```
getent hosts localhost
```
if it returns ::1, the kubectl port-forward will crash with the following error message:

```
W0331 09:47:04.409552   21488 portforward.go:165] Unable to listen on port {5001 5000}: Error parsing local part: strconv.ParseUint: parsing "": invalid syntax
```
This happens because of the returned address address from the Listen command used to bind the local port on "localhost" is then parsed by looking for a ":" char, which fails in case of the localhost is resolved to "::1"

@ncdc and others, could you please review ?

**Comment by akram** [comment]
I am the author of these commits

File: pkg/client/portforward/portforward_test.go [hunk 1]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
 	return http.Header{}
 }
 
+func TestGetListener(t *testing.T) {
+	stopChan := make(chan struct{}, 1)
+	pf, err := New(&client.Request{}, &client.Config{}, []string{"15000", "15000"}, stopChan)
+	if err != nil {
+		t.Logf("Cannot create PortForwarder %v: %v", pf, err)
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: This should be `t.Fatalf` as the remainder of the test code won't run without a PortForwarder

File: pkg/client/portforward/portforward_test.go [hunk 2]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+
+	}
+	upgrader := fakeUpgrader{conn: newFakeUpgradeConnection()}
+	pf.upgrader = &upgrader
+	testCases := []struct {
+		Hostname                  string
+		Protocol                  string
+		Ports                     ForwardedPort
+		ShouldRaiseError          bool
+		ExecpectedListenerAddress string
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: type: Expected (same for next line)

File: pkg/client/portforward/portforward_test.go [hunk 3]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+	if err != nil {
+		t.Logf("Cannot create PortForwarder %v: %v", pf, err)
+
+	}
+	upgrader := fakeUpgrader{conn: newFakeUpgradeConnection()}
+	pf.upgrader = &upgrader
+	testCases := []struct {
+		Hostname                  string
+		Protocol                  string
+		Ports                     ForwardedPort
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: I don't see any need to vary the ports in the test cases. I'd recommend hardcoding something inside the for loop instead.

File: pkg/client/portforward/portforward_test.go [hunk 4]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+		},
+		{
+			Hostname:         "[::1]",
+			Protocol:         "tcp4",
+			Ports:            ForwardedPort{16001, 16001},
+			ShouldRaiseError: true,
+		},
+
+		{
+			Hostname:         "www.google.com",
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: I don't really think you need this test case

File: pkg/client/portforward/portforward_test.go [hunk 5]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+			Protocol:         "tcp4",
+			Ports:            ForwardedPort{16002, 16002},
+			ShouldRaiseError: true,
+		},
+		{
+			Hostname:         "1.1.1.1",
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: I don't really think you need this test case

File: pkg/client/portforward/portforward_test.go [hunk 6]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+			ShouldRaiseError: true,
+		},
+	}
+
+	for i, testCase := range testCases {
+		listener, err := pf.getListener(testCase.Protocol, testCase.Hostname, &testCase.Ports)
+		errorRaised := err != nil
+
+		if testCase.ShouldRaiseError != errorRaised {
+			t.Fatalf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Do `t.Errorf` and `continue`

File: pkg/client/portforward/portforward_test.go [hunk 7]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+		}
+		if listener != nil {
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Convert to `if listener == nil { t.Errorf(); continue }` and remove the else

File: pkg/client/portforward/portforward_test.go [hunk 8]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+			host, port, err := net.SplitHostPort(listener.Addr().String())
+			t.Logf("Asked a %s forward for: %s:%v, got listener %s:%s", testCase.Protocol, testCase.Hostname, testCase.Ports, host, port)
+			if err != nil {
+				if testCase.ExpectedListenerAddress != host {
+					t.Fatalf("Listener does not listen on exepected address: asked %v got %v", testCase.ExpectedListenerAddress, host)
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Errorf

Add test case # in output

File: pkg/client/portforward/portforward_test.go [hunk 9]
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+				}
+				if testCase.ExpectedListenerPort != port {
+					t.Fatalf("Listener does not listen on exepected port: asked %v got %v", testCase.ExpectedListenerPort, port)
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Errorf

Add test case # in output

File: pkg/kubelet/dockertools/docker.go [hunk 10]
```diff
@@ -316,7 +316,11 @@ func (d *dockerContainerCommandRunner) PortForward(podInfraContainerID string, p
 	}
 
 	containerPid := container.State.Pid
-	// TODO use exec.LookPath for socat / what if the host doesn't have it???
+	// TODO what if the host doesn't have it???
+	path, lookupErr := exec.LookPath("socat")
+	if lookupErr != nil {
+		return fmt.Errorf("The socat command cannot be find in PATH=%s.", path)
```
Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: Just say "Port forwarding is unavailable: unable to locate socat"

File: pkg/kubelet/dockertools/docker.go [hunk 11]
```diff
@@ -316,7 +316,11 @@ func (d *dockerContainerCommandRunner) PortForward(podInfraContainerID string, p
 	}
 
 	containerPid := container.State.Pid
-	// TODO use exec.LookPath for socat / what if the host doesn't have it???
+	// TODO what if the host doesn't have it???
```
Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: Remove the TODO

**Comment by akram** [comment]
@ncdc thanks, for reviewing. I implemented your remarks.

 Team (@smarterclayton), could you please let me know your remarks?

For the moment, I will commit on top, to of course eventually squash and rebase the whole PR.

File: pkg/kubelet/dockertools/docker.go (hunk 11 above)
Comment (akram) [inline pkg/kubelet/dockertools/docker.go]: the question about what to do is still valid, for this moment, it is only failing with an error.
If you think that this is really sufficient, I can remove it.

File: pkg/kubelet/dockertools/docker.go (hunk 10 above)
Comment (akram) [inline pkg/kubelet/dockertools/docker.go]: More information isn't better than no information?

File: pkg/kubelet/dockertools/docker.go (hunk 11 above)
Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: I think the most appropriate way to handle this is to have the kubelet check for socat and return an error if it's not available when the port forwarding request comes in, before spawning listeners and so on. But that's not implemented now, and doesn't need to be part of this PR.

File: pkg/kubelet/dockertools/docker.go (hunk 10 above)
Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: I can't remember if this error is returned to the client, but if so, I don't think it's appropriate to tell the end user what the PATH is on the kubelet. You could certainly log this so it shows up in the kubelet's log file, but I don't think the end user should get the PATH info.

Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: And it's probably better not to mention socat either

File: pkg/client/portforward/portforward_test.go (hunk 4 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: This was one of the concerns of this PR.
Even even those tests are not 100% relevant for the current scenario, that helps to understand the expected behaviour.

More over, there are stil some stickyness to TCP4 in the docker.go implementation, once that we definitely want to get rid of this, and be able to handled tcp4 and tcp6 correctly, these tests would help.

If someone else agrees, I can remove these tests.

File: pkg/client/portforward/portforward_test.go (hunk 5 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: same comment as above :)

Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: I still disagree with including them. We're hardcoding localhost. I would actually modify `getListener` to not take in a hostname.

File: pkg/client/portforward/portforward.go [hunk 12]
```diff
@@ -205,6 +197,27 @@ func (pf *PortForwarder) listenOnPort(port *ForwardedPort) error {
 	return nil
 }
 
+// getListener creates a listener on the interface targeted by the given hostname on the given port with
+// the given protocol. protocol is in net.Listen style which basically admits values like tcp, tcp4, tcp6
+func (pf *PortForwarder) getListener(protocol string, hostname string, port *ForwardedPort) (net.Listener, error) {
```
Comment (ncdc) [inline pkg/client/portforward/portforward.go]: remove `hostname`

File: pkg/client/portforward/portforward.go [hunk 13]
```diff
@@ -205,6 +197,27 @@ func (pf *PortForwarder) listenOnPort(port *ForwardedPort) error {
...
+	listener, err := net.Listen(protocol, fmt.Sprintf("%s:%d", hostname, port.Local))
```
Comment (ncdc) [inline pkg/client/portforward/portforward.go]: remove `hostname` and hardcode `localhost`

Comment (ncdc) [inline pkg/client/portforward/portforward.go]: Oh I know why you're including it...

File: pkg/client/portforward/portforward_test.go (hunk 4 above)
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: The expected behavior is that we listen on localhost, not on google.com.

File: pkg/kubelet/dockertools/docker.go (hunk 10 above)
Comment (akram) [inline pkg/kubelet/dockertools/docker.go]: Agree with the PATH information not necessarily useful.

Regarding not mentioning socat: That would be very obscure. Not being able to know the cause of the  "Port forwarding is unavailable" is very frustrating. Why hiding it ?

Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: Reduce potential attack vectors from bad actors.

Comment (akram) [inline pkg/kubelet/dockertools/docker.go]: Security by obfuscation cannot be an argument....certainly in open source

Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: Alright

File: pkg/client/portforward/portforward_test.go (hunk 5 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: This is the purpose of this PR: To put in evidence the different behaviours of the net.Listen method depending on name resolution in the host.
These different behaviours are the cause of this issue in conjonction with the way we were parsing the resulting bind address.

Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: It's to make sure that we can get the port from "localhost:$port", where $port may be 0, in which case the listener's Addr will contain an arbitrary ephemeral port. Please remove tests for anything other than localhost an ::1.

File: pkg/client/portforward/portforward_test.go (hunk 6 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: will do that, commit is pending

File: pkg/client/portforward/portforward_test.go (hunk 7 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: will do that, commit is pending

File: pkg/client/portforward/portforward_test.go (hunk 8 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: will do that, commit is pending

File: pkg/client/portforward/portforward_test.go (hunk 9 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: will do that, commit is pending

File: pkg/kubelet/dockertools/docker.go (hunk 11 above)
Comment (akram) [inline pkg/kubelet/dockertools/docker.go]: agree
let's keep this for another PR for this specific point
I was only touching this part to remove frustration of not knowing the cause of the error.

**Comment by akram** [comment]
@ncdc do you see other points ?

File: pkg/client/portforward/portforward_test.go [hunk 14]
```diff
@@ -206,6 +206,86 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+	if err != nil {
+		t.Fatalf("Cannot create PortForwarder %v: %v", pf, err)
+
+	}
+	upgrader := fakeUpgrader{conn: newFakeUpgradeConnection()}
+	pf.upgrader = &upgrader
+	testCases := []struct {
+		Hostname                string
+		Protocol                string
+		Ports                   ForwardedPort
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: I still don't think you need this - just hardcode something at the beginning of the for loop as these values don't need to vary.

File: pkg/client/portforward/portforward_test.go [hunk 15]
```diff
@@ -206,6 +206,86 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+
+	for i, testCase := range testCases {
+		listener, err := pf.getListener(testCase.Protocol, testCase.Hostname, &testCase.Ports)
+		errorRaised := err != nil
+
+		if testCase.ShouldRaiseError != errorRaised {
+			t.Errorf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
+			continue
+		}
+		if listener != nil {
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]:
```go
if listener == nil {
  t.Errorf("...")
  continue
}

host, port, err ...
```

File: pkg/kubelet/dockertools/docker.go [hunk 16]
```diff
@@ -316,7 +316,11 @@ func (d *dockerContainerCommandRunner) PortForward(podInfraContainerID string, p
...
+	_, lookupErr := exec.LookPath("socat")
+	if lookupErr != nil {
+		return fmt.Errorf("The socat command cannot be find.")
```
Comment (ncdc) [inline pkg/kubelet/dockertools/docker.go]: Unable to do port forwarding: socat not found.

**Comment by bprashanth** [comment]
@ncdc i can't assign to you, so I'm assigning to Derek instead

**Comment by derekwaynecarr** [comment]
I am fine being @ncdc proxy on this, will wait for his LGTM

File: pkg/kubelet/dockertools/docker.go (hunk 16 above)
Comment (akram) [inline pkg/kubelet/dockertools/docker.go]: done, commit is pending

File: pkg/client/portforward/portforward_test.go (hunk 14 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: agree
I removed initialisation of pf. It is already tested elsewhere.
commit pending

File: pkg/client/portforward/portforward_test.go (hunk 15 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: Getting a nil listener is not an necessarely an erroneous test case.
We agree that the test case is considered in error if an error has not been raised where it should and vice-versa
And, when an error is not raised, the listener is not nil, and so, we have other tests then.

Or I misunderstood that. Failing when listener == nil definitely broke the test

File: pkg/client/portforward/portforward.go (hunk 13 above)
Comment (akram) [inline pkg/client/portforward/portforward.go]: So I can keep it? right?

Comment (ncdc) [inline pkg/client/portforward/portforward.go]: Yes

File: pkg/client/portforward/portforward_test.go (hunk 15 above)
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: If the test case should raise an error, and it did not, it will be caught by the previous `if` and then `continue`d. If the test case should not raise an error, the listener therefore should not be nil. If it is nil, this is an error.

File: pkg/client/portforward/portforward_test.go [hunk 17]
```diff
@@ -206,6 +206,85 @@ func (s *fakeUpgradeStream) Headers() http.Header {
 	return http.Header{}
 }
 
+func TestGetListener(t *testing.T) {
+	var pf PortForwarder
+	testCases := []struct {
+		Hostname                string
+		Protocol                string
+		Ports                   ForwardedPort
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Please remove this from the testCase struct.

File: pkg/client/portforward/portforward_test.go [hunk 18]
```diff
@@ -206,6 +206,85 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+		ShouldRaiseError        bool
+		ExpectedListenerAddress string
+		ExpectedListenerPort    string
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Please remove this from the testCase struct. You can hardcode a port below in the loop.

File: pkg/client/portforward/portforward_test.go (hunk 15 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: not exactly my comprehension:
OK, I got it, but, would the following be more accurate: 
If the test case dit not raise an error, the listener therefore should not be nil

I will commit this

File: pkg/client/portforward/portforward_test.go (hunk 17 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: OK
commit pending

File: pkg/client/portforward/portforward_test.go (hunk 18 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: OK commit pending

File: pkg/client/portforward/portforward_test.go (hunk 15 above)
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Whatever works for understanding it :smile: 

Basically if you don't expect an error, and you get a nil listener, that is a problem.

File: pkg/client/portforward/portforward_test.go [hunk 19]
```diff
@@ -206,6 +206,86 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+			ShouldRaiseError: true,
+		},
+	}
+
+	for i, testCase := range testCases {
+		expectedListenerPort := "12345"
+		listener, err := pf.getListener(testCase.Protocol, testCase.Hostname, &ForwardedPort{12345, 12345})
+		errorRaised := err != nil
+
+		if testCase.ShouldRaiseError != errorRaised {
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: I think this is cleaner

```go
if testCase.ShouldRaiseError != errorRaised {
	t.Errorf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
	continue
}
if errorRaised {
  continue
}

if listener == nil {
	t.Errorf("Test case #%d did not raised an error (%t) but failed in initializing listener", i, err)
	continue
}

host, port, _ := net.SplitHostPort(listener.Addr().String())
```

Comment (akram) [inline pkg/client/portforward/portforward_test.go]: ok,
commit pending

Comment (akram) [inline pkg/client/portforward/portforward_test.go]: committed

File: pkg/client/portforward/portforward_test.go [hunk 20]
```diff
@@ -206,6 +206,88 @@ func (s *fakeUpgradeStream) Headers() http.Header {
 	return http.Header{}
 }
 
+func CloseListener(l net.Listener) {
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: This method isn't used?

File: pkg/client/portforward/portforward_test.go [hunk 21]
```diff
@@ -206,6 +206,88 @@ func (s *fakeUpgradeStream) Headers() http.Header {
...
+		host, port, _ := net.SplitHostPort(listener.Addr().String())
+		t.Logf("Asked a %s forward for: %s:%v, got listener %s:%s, expected: %s", testCase.Protocol, testCase.Hostname, 12345, host, port, expectedListenerPort)
+		if host != testCase.ExpectedListenerAddress {
+			t.Errorf("Test case #%d failed: Listener does not listen on exepected address: asked %v got %v", i, testCase.ExpectedListenerAddress, host)
+		}
+		if port != expectedListenerPort {
+			t.Errorf("Test case #%d failed: Listener does not listen on exepected port: asked %v got %v", i, expectedListenerPort, port)
+
+		}
+		listener.Close()
```
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Sorry, I just noticed this - can you move this up to right before you call SplitHostPort and call it via defer?

File: pkg/client/portforward/portforward_test.go (hunk 20 above)
Comment (akram) [inline pkg/client/portforward/portforward_test.go]: oh, yes, sorry
Added it to properly handle closing in Code Review 5, forgot to remove it.
Will amend my last commit

File: pkg/client/portforward/portforward_test.go (hunk 21 above)
Comment (ncdc) [inline pkg/client/portforward/portforward_test.go]: Oh wait, that won't work because the same port is being reused.

**Comment by ncdc** [comment]
LGTM

@smarterclayton PTAL

**Comment by derekwaynecarr** [comment]
@akram - can you squash your commits down to 1?

**Comment by akram** [comment]
wow please wait, I don't know what happened when squashing my commits

**Comment by akram** [comment]
@derekwaynecarr done for squashing

**Comment by akram** [comment]
@derekwaynecarr :ping: squashed and rebased. Travis and shippable also seems ok.
